### PR TITLE
Default Perspectives in Pagination

### PR DESCRIPTION
**Pull request by BhaviKhator**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55257)

### CODE REVIEW

File: 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java [hunk 1]
```diff
@@ -524,7 +535,7 @@ public List<QLCEView> getAllViews(String accountId, String folderId, boolean inc
     sortCriteria = getModifiedSortCriteria(sortCriteria);
     Set<String> folderIds = new HashSet<>(Collections.singleton(folderId));
     List<CEView> viewList = ceViewDao.findByAccountIdAndFolderId(
-        accountId, folderIds, sortCriteria, pageNo, pageSize, searchKey, cloudFilters);
+        accountId, folderIds, sortCriteria, pageNo, pageSize, searchKey, cloudFilters, true);
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java:538]: Add pageNo is 0 check here as well.

File: 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/query/perspectives/PerspectivesQuery.java [hunk 2]
```diff
@@ -158,13 +158,13 @@ public PerspectiveData perspectives(@GraphQLArgument(name = "folderId") String f
       Set<String> allowedFolderIds = rbacHelper.checkFolderIdsGivenPermission(accountId, null, null,
           folders.stream().map(CEViewFolder::getUuid).collect(Collectors.toSet()), PERSPECTIVE_VIEW);
       perspectiveDataBuilder.totalCount(
-          ceViewService.countByAccountIdAndFolderId(accountId, allowedFolderIds, searchKey, cloudFilters));
+          ceViewService.countByAccountIdAndFolderId(accountId, allowedFolderIds, searchKey, cloudFilters,false));
       perspectiveDataBuilder.views(perspectiveService.perspectives(
```
Comment (sdsingh07) [inline 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/query/perspectives/PerspectivesQuery.java:162]: Create new overloaded function for REST API in PerspectiveService and CEViewService.

File: 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java [hunk 3]
```diff
@@ -261,13 +261,29 @@ public long count(String accountId) {
...
   }
 
-  public Query<CEView> paginationQuery(
+  public Query<CEView> getDefaultFilterCount(Query<CEView> query, boolean excludeDefault) {
+    if (excludeDefault) {
+      query.filter(CEViewKeys.viewType, ViewType.DEFAULT);
+    }
+    return query;
+  }
+  public Query<CEView> getDefaultFilterPerspectives(Query<CEView> query, boolean excludeDefault) {
```
[the remark this answers is not in the document]
Comment (sdsingh07) [reply 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java:271]: Delete this function.

File: 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java [hunk 4]
```diff
@@ -295,13 +305,24 @@ public Query<CEView> paginationQuery(
     return query;
   }
 
+  public List<CEView> findByAccountIdAndFolderId(String accountId, Set<String> folderIds,
+      QLCEViewSortCriteria sortCriteria, Integer pageNo, Integer pageSize, String searchKey, List<CloudFilter> filters,
+      boolean excludeDefault, Long defaultPerspectiveCount) {
+    Query<CEView> query = getPaginationFiltersQuery(accountId, folderIds, searchKey, filters);
+    query = getDefaultFilterPerspectives(query, excludeDefault);
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java:312]: Add check directly here. 

```
if (excludeDefault) { query.field(CEViewKeys.viewType).notEqual(ViewType.DEFAULT); }
```

File: 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java [hunk 5]
```diff
@@ -295,13 +305,24 @@ public Query<CEView> paginationQuery(
...
+    query = decorateQueryWithSortCriteria(query, sortCriteria);
+    int offset = Objects.equals(pageNo, 0) ? 0 : (int) (pageNo * pageSize - defaultPerspectiveCount);
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java:314]: Move this logic to service level. Pass limit and offset as arguments, not pageSize and pageNo.

File: 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java [hunk 6]
```diff
@@ -295,13 +305,24 @@ public Query<CEView> paginationQuery(
...
+  }
   public List<CEView> findByAccountIdAndFolderId(String accountId, Set<String> folderIds,
       QLCEViewSortCriteria sortCriteria, Integer pageNo, Integer pageSize, String searchKey,
       List<CloudFilter> filters) {
-    Query<CEView> query = paginationQuery(accountId, folderIds, searchKey, filters);
+    Query<CEView> query = getPaginationFiltersQuery(accountId, folderIds, searchKey, filters);
     query = decorateQueryWithSortCriteria(query, sortCriteria);
-    query.offset(pageNo * pageSize);
     query.limit(pageSize);
+    query.offset(pageNo * pageSize);
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java:325]: Pass limit and offset as arguments, not pageSize and pageNo.

File: 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java [hunk 7]
```diff
@@ -295,13 +305,24 @@ public Query<CEView> paginationQuery(
...
+      QLCEViewSortCriteria sortCriteria, Integer pageNo, Integer pageSize, String searchKey, List<CloudFilter> filters,
+      boolean excludeDefault, Long defaultPerspectiveCount) {
+    Query<CEView> query = getPaginationFiltersQuery(accountId, folderIds, searchKey, filters);
+    query = getDefaultFilterPerspectives(query, excludeDefault);
+    query = decorateQueryWithSortCriteria(query, sortCriteria);
+    int offset = Objects.equals(pageNo, 0) ? 0 : (int) (pageNo * pageSize - defaultPerspectiveCount);
+    query.limit(pageSize);
+    query.offset(offset);
+    return query.asList();
+  }
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java:318]: Add new line.

File: 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java [hunk 8]
```diff
@@ -295,13 +305,24 @@ public Query<CEView> paginationQuery(
...
   public List<CEView> findByAccountIdAndFolderId(String accountId, Set<String> folderIds,
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java:319]: Remove this function. There should be only one `findByAccountIdAndFolderId` function.

File: 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java [hunk 9]
```diff
@@ -295,13 +305,24 @@ public Query<CEView> paginationQuery(
     return query;
   }
 
+  public List<CEView> findByAccountIdAndFolderId(String accountId, Set<String> folderIds,
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java:308]: Rename it to `findByAccountIdAndFolderIds`

File: 450-ce-views/src/main/java/io/harness/ccm/views/service/CEViewService.java [hunk 10]
```diff
@@ -46,10 +46,16 @@ List<QLCEView> getAllViews(
...
       Integer pageSize, Integer pageNo, String searchKey, List<CEViewFolder> folders, Set<String> allowedFolderIds,
       List<CloudFilter> cloudFilters);
+  List<QLCEView> getAllViews(String accountId, boolean includeDefault, QLCEViewSortCriteria sortCriteria,
+      Integer pageSize, Integer pageNo, String searchKey, List<CEViewFolder> folders, Set<String> allowedFolderIds,
+      List<CloudFilter> cloudFilters, boolean excludeDefault);
   List<QLCEView> getAllViews(String accountId, String folderId, boolean includeDefault,
       QLCEViewSortCriteria sortCriteria, Integer pageSize, Integer pageNo, String searchKey,
       List<CloudFilter> cloudFilters);
 
+  public List<CEView> getFirstPageViews(String accountId, QLCEViewSortCriteria sortCriteria, Integer pageSize,
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/service/CEViewService.java:56]: Remove this function.

File: 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java [hunk 11]
```diff
@@ -507,11 +507,42 @@ public List<CEView> getAllViews(String accountId) {
     return ceViewDao.list(accountId);
   }
 
+  @Override
+  public List<CEView> getFirstPageViews(String accountId, QLCEViewSortCriteria sortCriteria, Integer pageSize,
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java:511]: Make this function private and remove Override annotation.

File: 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java [hunk 12]
```diff
@@ -507,11 +507,42 @@ public List<CEView> getAllViews(String accountId) {
...
+      Integer pageNo, String searchKey, Set<String> allowedFolderIds, List<CloudFilter> cloudFilters) {
+    List<CEView> defaultViewList = ceViewDao.findByAccountIdAndFolderId(
+        accountId, allowedFolderIds, sortCriteria, pageNo, pageSize, searchKey, cloudFilters, false, 0L);
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java:514]: Pass limit and offset only. Remove defaultPerspCount as argument.

File: ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/perspectives/PerspectiveResource.java [hunk 13]
```diff
@@ -462,7 +462,7 @@ private void checkPerspectiveExists(String perspectiveId, CEView ceView) {
     perspectiveDataBuilder.totalCount(
         ceViewService.countByAccountIdAndFolderId(accountId, allowedFolderIds, searchKey, cloudFilters));
     perspectiveDataBuilder.views(perspectiveService.perspectives(
-        sortCriteria, accountId, pageSize, pageNo, searchKey, folders, allowedFolderIds, cloudFilters));
+        sortCriteria, accountId, pageSize, pageNo, searchKey, folders, allowedFolderIds, cloudFilters, true));
```
Comment (sdsingh07) [inline ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/perspectives/PerspectiveResource.java:465]: excludeDefault as false. And, test this API as well.

File: 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java [hunk 14]
```diff
@@ -507,11 +507,42 @@ public List<CEView> getAllViews(String accountId) {
...
+    Long defaultPerspectiveCount = (long) defaultViewList.size();
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java:515]: defaultPerspectiveCount should be of `int` type.

File: 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java [hunk 15]
```diff
@@ -507,11 +507,42 @@ public List<CEView> getAllViews(String accountId) {
...
   public List<QLCEView> getAllViews(String accountId, boolean includeDefault, QLCEViewSortCriteria sortCriteria,
       Integer pageSize, Integer pageNo, String searchKey, List<CEViewFolder> folders, Set<String> allowedFolderIds,
       List<CloudFilter> cloudFilters) {
     sortCriteria = getModifiedSortCriteria(sortCriteria);
+    List<CEView> viewList;
+    if (Objects.equals(pageNo, 0)) {
+      viewList =
+          getFirstPageViews(accountId, sortCriteria, pageSize, pageNo, searchKey, allowedFolderIds, cloudFilters);
+    } else {
+      Long defaultPerspectiveCount = countByAccountIdAndFolderId(
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java:532]: Remove this call. If page is other than 0, defaultPerspectiveCount will always be 0.

File: 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java [hunk 16]
```diff
@@ -523,8 +554,15 @@ public List<QLCEView> getAllViews(String accountId, String folderId, boolean inc
       List<CloudFilter> cloudFilters) {
     sortCriteria = getModifiedSortCriteria(sortCriteria);
     Set<String> folderIds = new HashSet<>(Collections.singleton(folderId));
-    List<CEView> viewList = ceViewDao.findByAccountIdAndFolderId(
-        accountId, folderIds, sortCriteria, pageNo, pageSize, searchKey, cloudFilters);
+    List<CEView> viewList;
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java:557]: Extract the new code to common function and use it. Currently, we are duplicating this code at 2 places.

File: 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java [hunk 17]
```diff
@@ -523,8 +554,15 @@ public List<QLCEView> getAllViews(String accountId, String folderId, boolean inc
...
+    if (Objects.equals(pageNo, 0)) {
+      viewList = getFirstPageViews(accountId, sortCriteria, pageSize, pageNo, searchKey, folderIds, cloudFilters);
+    } else {
+      Long defaultPerspectiveCount =
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java:561]: Remove this call. If page is other than 0, defaultPerspectiveCount will always be 0.

File: 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java [hunk 18]
```diff
@@ -523,8 +554,15 @@ public List<QLCEView> getAllViews(String accountId, String folderId, boolean inc
...
+          countByAccountIdAndFolderId(accountId, folderIds, searchKey, Collections.singletonList(CloudFilter.DEFAULT));
+      viewList = ceViewDao.findByAccountIdAndFolderId(
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java:563]: Calculate limit and offset here only and pass them to the Dao layer.

File: 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java (hunk 17 above)
Comment (BhaviKhator) [reply 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java:561]: need it to set offset

File: 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java (hunk 15 above)
Comment (BhaviKhator) [reply 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java:532]: need it for setting offset for next pages

Comment (sdsingh07) [reply 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java:532]: Calculate limit and offset at service level.

File: 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java (hunk 17 above)
Comment (sdsingh07) [reply 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewServiceImpl.java:561]: Calculate limit and offset at service level.

File: 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java (hunk 8 above)
Comment (sdsingh07) [reply 450-ce-views/src/main/java/io/harness/ccm/views/dao/CEViewDao.java:319]: We need this function, my bad.

**Comment by sdsingh07** [comment]
New PR: https://github.com/harness/harness-core/pull/55301

**Pull request closed without merging**